cpp/src/arrow/csv/writer.cc

```diff
@@ -0,0 +1,398 @@
+    populator = new QuotedColumnPopulator(pool, end_char);
+    return Status::OK();
+  }
+
+  template <typename TypeClass>
+  enable_if_dictionary<TypeClass, Status> Visit(const TypeClass& type) {
+    return VisitTypeInline(*type.value_type(), this);
+  }
+
+  template <typename TypeClass>
+  enable_if_t<is_nested_type<TypeClass>::value || is_extension_type<TypeClass>::value,
+              Status>
+  Visit(const TypeClass& type) {
+    return Status::Invalid("Nested and extension types not supported");
+  }
+
+  template <typename TypeClass>
+  enable_if_t<is_primitive_ctype<TypeClass>::value || is_decimal_type<TypeClass>::value ||
+                  is_null_type<TypeClass>::value || is_temporal_type<TypeClass>::value,
+              Status>
+  Visit(const TypeClass& type) {
+    populator = new UnquotedColumnPopulator(pool, end_char);
+    return Status::OK();
+  }
+
+  char end_char;
+  MemoryPool* pool;
+  ColumnPopulator* populator;
+};
+
+Result<std::unique_ptr<ColumnPopulator>> MakePopulator(const Field& field, char end_char,
+                                                       MemoryPool* pool) {
+  PopulatorFactory factory{end_char, pool, nullptr};
+  RETURN_NOT_OK(VisitTypeInline(*field.type(), &factory));
+  return std::unique_ptr<ColumnPopulator>(factory.populator);
+}
+
+class CsvConverter {
+ public:
+  static Result<std::unique_ptr<CsvConverter>> Make(std::shared_ptr<Schema> schema,
+                                                    MemoryPool* pool) {
+    std::vector<std::unique_ptr<ColumnPopulator>> populators(schema->num_fields());
+    for (int col = 0; col < schema->num_fields(); col++) {
+      char end_char = col < schema->num_fields() - 1 ? ',' : '\n';
+      ASSIGN_OR_RAISE(populators[col],
+                      MakePopulator(*schema->field(col), end_char, pool));
+    }
+    return std::unique_ptr<CsvConverter>(
+        new CsvConverter(std::move(schema), std::move(populators), pool));
+  }
+  static constexpr int64_t kColumnSizeGuess = 8;
+  Status WriteCsv(const RecordBatch& batch, const WriteOptions& options,
+                  io::OutputStream* out) {
+    RETURN_NOT_OK(PrepareForContentsWrite(options, out));
+    RecordBatchIterator iterator = batch.SliceIterator(options.batch_size);
+    for (auto maybe_slice : iterator) {
+      ASSIGN_OR_RAISE(std::shared_ptr<RecordBatch> slice, maybe_slice);
+      RETURN_NOT_OK(TranslateMininalBatch(*slice));
+      RETURN_NOT_OK(out->Write(data_buffer_));
+    }
+    return Status::OK();
+  }
+
+  Status WriteCsv(const Table& table, const WriteOptions& options,
+                  io::OutputStream* out) {
+    TableBatchReader reader(table);
+    reader.set_chunksize(options.batch_size);
+    RETURN_NOT_OK(PrepareForContentsWrite(options, out));
+    std::shared_ptr<RecordBatch> batch;
+    RETURN_NOT_OK(reader.ReadNext(&batch));
+    while (batch != nullptr) {
+      RETURN_NOT_OK(TranslateMininalBatch(*batch));
+      RETURN_NOT_OK(out->Write(data_buffer_));
```

Comment:
Another easy optimization is to move the writes to a dedicated background thread.  I'll add this to my list of "to asyncerize" which will include this optimization by necessity.

Reply:
Yes, agreed.  My initial use-case for this will be to a bufferoutputstream so asynchronous isn't important for that at least.  This actually raises a question on whether detecting blocking vs non-blocking IO sources is important for threading.

Reply:
I don't know that a buffered output stream solves the problem I'm worried about.  It will help mitigate the cost of many small writes by grouping them but the large writes still take time.  So if the CSV is large enough to span multiple buffers you still have to block here occasionally.  I may be misunderstanding what you mean by bufferedoutputstream though.

The detecting question is a good one.  I think it boils down like so...

Currently, we assume all streams are blocking (buffered output stream is still a blocking stream, it just blocks less).  This means that anytime we do I/O from an async context we need to "background" it by creating a dedicated thread to do the I/O (very soon I hope this will switch to "borrowing a dedicated thread from the I/O pool")

At some point, as more underlying filesystems are non-blocking, we can reverse the assumption and assume that all underlying streams are non-blocking.  Any remaining blocking streams will then be responsible for the "background"ing.

Reply:
> I don't know that a buffered output stream solves the problem I'm worried about. It will help mitigate the cost of many small writes by grouping them but the large writes still take time. So if the CSV is large enough to span multiple buffers you still have to block here occasionally. I may be misunderstanding what you mean by bufferedoutputstream though.

[BufferOutputStream](https://github.com/apache/arrow/blob/57376d28cf433bed95f19fa44c1e90a780ba54e8/cpp/src/arrow/io/memory.h#L39) is an in memory output stream that writes to a resizable buffer.

Reply:
I'm not sure why we're tallking about doing writes in a separate thread. Writes are typically asynchronous, so the only cost is a memory copy (and perhaps a system call).

Reply:
Ok, yes, not a problem at all for `BufferOutputStream`.  That will teach me to read more closely.

@pitrou You are correct.  Simply adding a background readahead style thread will just introduce a second cache in addition to the OS' existing cache (or whatever dirty page cache is in our s3 filesystem) and that doesn't help anything.  I still think we will eventually want to make writes properly async (futures).  In the event we are in a huge write or writing to a slow (S3) source we still might block, although the gain there is not performance of this CSV write (the one blocked by I/O) as it is keeping the CPU threads clear in case other tasks are going on at the same time.